File: js/word_290.js
var major = (pattern, s) => {
  let sArr = s.split(" ");

  if (sArr.length != pattern.length) {
    return false;
  }

  let map = new Map();
  let set = new Set();

  for (let i = 0; i < pattern.length; i++) {
    if (map.get(pattern[i]) === undefined) {
      if (set.has(sArr[i])) {
        return false;
      }

      map.set(pattern[i], sArr[i]);
      set.add(sArr[i]);
    } else {
      if (map.get(pattern[i]) !== sArr[i]) {
        return false;
      }
    }
  }

  return true;
};

const resp = major("ab", "dog dog");
console.log("resp", resp);

File: dart/word_290.dart
void main() {
  final resp = wordPattern("ab", "dog dog");
  print(resp);
}

bool wordPattern(String pattern, String s) {
  List<String> pArr = pattern.split("");
  List<String> sArr = s.split(" ");

  Map<String, String> map = {};
  Set<String> set = Set();

  for (int i = 0; i < pArr.length; i++) {
    if (map.containsKey(pArr[i])) {
      if (map[pArr[i]] != sArr[i]) {
        return false;
      }
    } else {
      if (set.contains(sArr[i])) {
        return false;
      }

      map[pArr[i]] = sArr[i];
      set.add(sArr[i]);
    }
  }

  return true;
}
